clients/client-py/taskcluster/aio/upload.py

```diff
@@ -146,10 +147,38 @@ async def reader_gen():
     resp.raise_for_status()
 
 
-class HashingReader(SyncHashingReader):
-    """An async version of the sync HashingReader"""
+class HashingReader:
+    """A Reader implementation that hashes contents as they are read."""
+
+    def __init__(self, inner):
+        self.inner = inner
+        self.sha256 = hashlib.sha256()
+        self.sha512 = hashlib.sha512()
+        self.bytes = 0
 
     async def read(self, max_size):
-        chunk = await self.inner.read(max_size)
-        self.update(chunk)
+        chunk = await ensureCoro(self.inner.read)(max_size)
+        try:
+            self.update(chunk)
+        except TypeError:
+            # Best effort: since ensureCoro fails to identify certain cases as
+            # coroutines (e.g. with aiofiles), when we fail to get a bytes-like object
+            # on the first try, we retry without the ensureCoro wrapping.
+            chunk = await self.inner.read(max_size)
+            self.update(chunk)
         return chunk
```

Comment:
This concerns me, since it's calling `read(..)` twice, and that function has side-effects (incrementing the current position).

We actually know the cases where passed functions are sync, so maybe we could do this wrapping explicitly at that point, in the sync versions -- at least for the readerFactory and writerFactories.